src/Workspaces/Core/Portable/FlowAnalysis/SymbolUsageAnalysis/SymbolUsageAnalysis.BasicBlockAnalysisData.cs

```diff
@@ -0,0 +1,178 @@
+            {
+                FreeAndClearValues();
+                s_pool.Free(this);
+            }
+
+            private void FreeAndClearValues()
+            {
+                foreach (var value in _reachingWrites.Values)
+                {
+                    value.Free();
+                }
+
+                _reachingWrites.Clear();
+            }
+
+            public void SetAnalysisDataFrom(BasicBlockAnalysisData other)
+            {
+                if (ReferenceEquals(this, other))
+                {
+                    return;
+                }
+
+                FreeAndClearValues();
+                AddEntries(_reachingWrites, other);
+            }
+
+            /// <summary>
+            /// Gets the currently reachable writes for the given symbol.
+            /// </summary>
+            public IEnumerable<IOperation> GetCurrentWrites(ISymbol symbol)
+            {
+                if (_reachingWrites.TryGetValue(symbol, out var values))
+                {
+                    foreach (var value in values)
+                    {
+                        yield return value;
+                    }
+                }
+            }
+
+            /// <summary>
+            /// Marks the given symbol write as a new unread write operation,
+            /// potentially clearing out the prior write operations if <paramref name="maybeWritten"/> is <code>false</code>.
+            /// </summary>
+            public void OnWriteReferenceFound(ISymbol symbol, IOperation operation, bool maybeWritten)
+            {
+                if (!_reachingWrites.TryGetValue(symbol, out var values))
+                {
+                    values = PooledHashSet<IOperation>.GetInstance();
+                    _reachingWrites.Add(symbol, values);
+                }
+                else if (!maybeWritten)
+                {
+                    values.Clear();
+                }
+
+                values.Add(operation);
+            }
+
+            public bool Equals(BasicBlockAnalysisData other)
+                => this.GetHashCode() == (other?.GetHashCode() ?? 0);
+
+            public override bool Equals(object obj)
+            {
+                return Equals(obj as BasicBlockAnalysisData);
+            }
+
+            public override int GetHashCode()
+            {
+                var hashCode = _reachingWrites.Count;
+                foreach ((int keyHash, int valueHash) in _reachingWrites.Select(GetKeyValueHashCodeTuple).OrderBy(hashTuple => hashTuple.keyHash))
+                {
+                    hashCode = Hash.Combine(Hash.Combine(keyHash, valueHash), hashCode);
+                }
+
+                return hashCode;
+
+                // Local functions.
+                (int keyHash, int valueHash) GetKeyValueHashCodeTuple(KeyValuePair<ISymbol, PooledHashSet<IOperation>> keyValuePair)
+                    => (keyValuePair.Key.GetHashCode(), Hash.Combine(keyValuePair.Value.Count, Hash.CombineValues(keyValuePair.Value)));
```

Comment:
nit, consider extracting to ensure no accidental capture.  (note: this concern goes away once we have static-local-funcitons).